### PR TITLE
Enhance SetupConnection Handling and Implement Job declaration Flag Checks 

### LINKED ISSUE #853: Issue with flags checking and related roles behaviour
In this [todo!](https://github.com/stratum-mining/stratum/blob/dev/protocols/v2/subprotocols/common-messages/src/setup_connection.rs#L88) we see that the check_flags panics if used in any other protocol except `MininProtocol`. Indeed, this method is used in `IsUpstream` trait, that does not have an extension (possibly empty) to the `JobDeclaration` protocol which is implemented for JDS
As a result, apparently there is no way for the JDS to check the bit `REQUIRES_ASYNC_JOB_MINING` of `SetupConnection.flag` in JobDeclaration protocol, making the SRI slightly no specs compliant.

1. Things to do: test with MG if different flags trigger the desired behaviour in the particular case above.
2. Check this issue affects also other parts of the code.

### PR DESCRIPTION
**Pull request by Shourya742**
This PR closes #853:

1. Completed MG Test.
2. Add flag check during SetupConnection call.

### CODE REVIEW

**Comment by jbesraa** [comment]
@Shourya742 could you please tide up the commit history?

**Comment by Shourya742** [comment]
> @Shourya742 could you please tide up the commit history?

Ah, I'm not sure if we should do that. This PR is based on the remaining tasks left from one of the stale PRs (#918). I'm hesitant to undermine the previous contributor's commit.

**Comment by Shourya742** [comment]
@jbesraa  made the commits modular and relevant to the context. Please review and suggest any changes.

**Review by jbesraa — COMMENTED** [review]
Looks good. Just added two nits.
Could you also please take out the `Cargo.lock` file changes?

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (Shourya742) [reply roles/jd-server/src/lib/job_declarator/mod.rs:456]: These unwraps are safe, and since we are the ones creating the response, I don't think there is a need for additional error handling as this will always succeed. However, I can add an expect block for better readability in case of an error.

File: test/message-generator/mock/job-declarator-mock-invalid-flag.json
Comment (Shourya742) [reply test/message-generator/mock/job-declarator-mock-invalid-flag.json]: Yes, it should be. Can we let it go this time?

File: test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json
Comment (plebhash) [inline test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json:22]: hmm are you repurposing the config file from a different test?

I'd lean towards copying this directory into a new one with a more meaningful name. Even if the contents are identical and it ends up being somewhat redundant, if feels kinda dirty to have something here that is completely unrelated to the test.

File: test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json
Comment (plebhash) [inline test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json:1]: `job-declarator` used to be the old name for JDS, before we introduced the conceptual distinction between JDS vs JDC

I would avoid using this generic naming, even if we have some legacy tests that still follow this convention

it seems this test is focusing on JDS, so we should call it `jds-*`

File: test/message-generator/mock/job-declarator-mock-invalid-flag.json
Comment (plebhash) [inline test/message-generator/mock/job-declarator-mock-invalid-flag.json:5]:
```suggestion
        "Soft mock of JDC",
```

File: test/message-generator/mock/job-declarator-mock-invalid-flag.json
Comment (plebhash) [inline test/message-generator/mock/job-declarator-mock-invalid-flag.json:1]: this comment is similar to what I said about the `job-declarator-flag-test.json` file

I'd rename this file to ``jdc-mock-for-jds-setupconnection-flag-test.json``

File: test/message-generator/messages/common_messages.json
Comment (plebhash) [inline test/message-generator/messages/common_messages.json:65]: can you explain the rationale on this?
why is `"flags": 6`?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:100]: why is `Protocol::JobDistributionProtocol` on this line?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:101]: @Fi3 can you confirm Template Distribution Protocol does not set any flags for `SetupConnection`?

I cannot find anything on [the specs](https://stratumprotocol.org/specification/07-Template-Distribution-Protocol/) but just to make sure.

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:409]: it seems we're changing this line just to introduce a typo?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:413]: it seems we're changing this line just to introduce a typo?

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [inline roles/jd-server/src/lib/job_declarator/mod.rs:476]: shouldn't we do some kind of verification to make sure this is indeed a `SetupConnection` that JDS just received?

what if the client sent some nonsense and we're blindly parsing this payload as if it was a `SetupConnection` flag?

Comment (plebhash) [reply roles/jd-server/src/lib/job_declarator/mod.rs:476]: also, what is the rationale behind `payload[5]`?

**Review by plebhash — CHANGES_REQUESTED** [review]
this PR has a solid foundation and demonstrates some great work by @Shourya742 

but we still need to improve some things before we are able to merge it

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (Shourya742) [reply roles/jd-server/src/lib/job_declarator/mod.rs:476]: I completely overlooked this possibility. It makes sense to have this approach rather than parsing based on my understanding. The reason for payload[5] was that the flag is a 4-byte value encoded in the payload in little-endian format, so the 5th byte must contain the flag value. Since the current range for the flag value is up to 3 bits max, it fits easily within 8 bits or the 1st byte. However, we still have the possibility of increasing the flag bit range in the future.

For any other type of message except the setupConnection, we should not respond. Instead, we should continue listening to our other processes, as these messages might be some sort of attack to slow down our JDS server. In such cases, we should avoid sending any response to the connection request.

But ya thanks for pointing this out. @plebhash

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Shourya742) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:413]: Will correct it out

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Shourya742) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:409]: will correct it out

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Shourya742) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:100]: Because for both of these cases, we dont receive any flags so no point of checking for them

File: test/message-generator/messages/common_messages.json
Comment (Shourya742) [reply test/message-generator/messages/common_messages.json:65]: Okay, so the idea is that a setupConnection request from JDC to JDS needs to have a flag with the 0th bit set. This flag value have both the 1st and 2nd bits set but not the 0th. In such a case, the expected behavior is a setupConnection error.

File: test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json
Comment (Shourya742) [reply test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json:1]: Yup makes sense

File: test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json
Comment (Shourya742) [reply test/message-generator/test/job-declarator-flag-test/job-declarator-flag-test.json:22]: Okay, I will create a separate configuration file. It makes sense to have different configuration files for different contexts.

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Fi3) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:101]: I confirm

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [reply roles/jd-server/src/lib/job_declarator/mod.rs]: > The reason for payload[5] was that the flag is a 4-byte value encoded in the payload in little-endian format, so the 5th byte must contain the flag value.

we have APIs for parsing messages from frame payloads, so we should use them for safety and readability

> For any other type of message except the setupConnection, we should not respond. Instead, we should continue listening to our other processes, as these messages might be some sort of attack to slow down our JDS server. In such cases, we should avoid sending any response to the connection request.

the attack vector is a good point that I also overlooked in my previous message.

given that the context of this code is a `while let Ok((stream, _)) = listner.accept().await { ... }`, I wonder if it doesn't make sense to just drop the connection for every case where the first incoming message is not a `SetupConnection`?

cc @Fi3

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:100]: ah ok, for some reason my brain parsed `Protocol::JobDeclarationProtocol` and `Protocol::JobDistributionProtocol`  as if they were the same thing

which is actually a good opportunity to highlight a potential footgun:

I'm about 95% sure that `JobDistributionProtocol` is just some legacy terminology that we forgot to remove after some spec changes

if we check [stratumprotocol.org/specification](https://stratumprotocol.org/specification/) we'll only find 3 subprotocols:
- Mining
- Job Declaration
- Template Distribution

@Fi3 do you confirm that we should clean up these remaining `JobDistributionProtocol` from the code?

if that's the case, we should create an issue so we can clean this up in the future

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:101]: ok so @Shourya742 we should change this comment

instead of saying we are assuming, we should confidently state this scenario does not have any meaningful flags for `SetupConnection` message

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Fi3) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:100]: if we are not going to add it the spec I would remove it

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (lorbax) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:96]: I don't understand semantic of this.
You just have to check the 31st bit, why you do the & bitwise between the whole `available` and `required`?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (lorbax) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:95]: seems that this produces false when the async mining job is available in the upstream, but it is not required by the downstream. In my opinion this should produce true, not false. 
Also, this produces true when async mining job is required by the downstrem bu not available by the upstream. I think that this should produce false.
Passed  $  Self  $  !Self || Passed
      T       $    F    $            T
      T       $    T    $            T
      F       $    T    $            F
      F       $    F    $            T

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (lorbax) [inline roles/jd-server/src/lib/job_declarator/mod.rs:479]: here the JDS offers async job mining, so if the downstream does not requires it, the flags checking should produce `True`.
Aslo, according to the fact that the client set or not the flag `async mining job`, we espect different behaviours from the JDS. Perhaps this is material for a second PR.

File: test/message-generator/messages/common_messages.json
Comment (plebhash) [reply test/message-generator/messages/common_messages.json:65]: why not simply `0`?

1st and 2nd bits will be shifted away and ignored inside `check_flags`, so it feels like there's no real upside for using `6` here, it just makes it more confusing to understand

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [inline roles/jd-server/src/lib/job_declarator/mod.rs:484]: I would change this line, and simply do a `if is_valid`

then, declare `sv2_frame: StdFrame` inside the `{}` block, and also move the calls to `JobDeclaratorDownstream::new` + `JobDeclaratorDownstream::start` from line 514 to here, while doing the same for the `else` case

this way, we can avoid checking `is_valid` 2x, which feels a bit redundant

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [reply roles/jd-server/src/lib/job_declarator/mod.rs]: I'm not a fan of `unwrap` either, but there's many instances where getting rid of them requires changing breaking APIs, which drastically change the nature of the task at hand

this kind of thing should be addressed during the refactor efforts

for now, I think replacing `unwrap` with `expect` where possible feels like a good start

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:95]: @lorbax has a point, here's a refined truth table:
| available | required | check |
|-----------|----------|-------|
| T         | T        | T     |
| T         | F        | T     |
| F         | T        | F    |
| F         | F        | T     |

I'm also writing some suggestions on the code

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [inline roles/jd-server/src/lib/job_declarator/mod.rs:482]: it seems we were mixing up the order of the parameters:

second parameter is `available_flags`, which signals that JDS supports async jobs, and therefore should be `1`

third parameter is `required_flags`,  which signals if JDC requested async jobs or not, so that one should be `flag`
```suggestion
                            let is_valid = SetupConnection::check_flags(
                                Protocol::JobDeclarationProtocol,
                                flag,
                                1,
                            );
```

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:95]:
```suggestion
                match (requires_async_job_mining_self, requires_async_job_mining_passed) {
                    (true, true) => true,
                    (true, false) => true,
                    (false, true) => false,
                    (false, false) => true,
                }
```

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:96]: I agree this line is unnecessary, and should be replaced by the `match` suggestion above
```suggestion
```

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (plebhash) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:98]: the `match` suggestion above already returns the `bool`
```suggestion
```

**Comment by plebhash** [comment]
@lorbax provided some valuable feedback https://github.com/stratum-mining/stratum/pull/1035#discussion_r1679624474 where basically @Shourya742 and I were missing the fact that some deeper changes to JDS were actually necessary

if we carefully read #853, a logical conclusion is that JDS needs a new config parameter, where the user chooses if JDS should support async jobs or not

that was achieved on this commit: 996adc37532d246c8d2594360c7c8b9eacdf56ac

---

also, we need MG tests that will cover this entire truth table:
| async jobs allowed | async jobs requested | `SetupConnection.Success` |
|-----------|----------|-------|
| T         | T        | T     |
| T         | F        | T     |
| F         | T        | F    |
| F         | F        | T     |

that was achieved on this commit: 35b1e30ded5f78760e5fea37d155e2c3a815e5ff

---

I suggest @Shourya742 reviews these two commits, and if he agrees, cherry-pick them into this PR

then, @lorbax can verify if now we finally have a sane solution to #853

File: roles/jd-server/src/lib/job_declarator/mod.rs
Comment (plebhash) [reply roles/jd-server/src/lib/job_declarator/mod.rs]: you are right!

and I think we should fix that on this PR, otherwise this PR would be introducing a broken feature into JDS

I elaborate more here https://github.com/stratum-mining/stratum/pull/1035#issuecomment-2237516417

**Review by plebhash — APPROVED** [review]

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (jbesraa) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:104]: @Shourya742 I think as well this shouldn't be part of this PR, or at least it should in its own commits explaining what/why

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (jbesraa) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:72]: Could you please add an explanation above the L72 describing what are the checks we are doing?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (jbesraa) [inline protocols/v2/subprotocols/common-messages/src/setup_connection.rs:88]: I think here as well(above L88) it would be useful to add explaination

**Review by jbesraa — CHANGES_REQUESTED** [review]
This looks very close. Added few comments.
Could you please rearrange the commits as well?

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (Shourya742) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:72]: Aren't these self-explanatory, as in taking the first bit and checking the condition based on whether it's set or not? However, I will still add documentation if it's unclear to others as well.

**Comment by Shourya742** [comment]
> This looks very close. Added few comments. Could you please rearrange the commits as well?

Yes, I'll squash a few of my commits and add descriptions according to the context

File: protocols/v2/subprotocols/common-messages/src/setup_connection.rs
Comment (jbesraa) [reply protocols/v2/subprotocols/common-messages/src/setup_connection.rs:72]: Yea, I think it would be better to document those. Not necessarily line by line, be the general approach

**Comment by jbesraa** [comment]
> > This looks very close. Added few comments. Could you please rearrange the commits as well?
> 
> Yes, I'll squash a few of my commits and add descriptions according to the context

Thanks, this might be helpful for when writing commits  https://cbea.ms/git-commit/

**Comment by Shourya742** [comment]
@jbesraa I’ve added the suggested changes. Please review and let me know if any further modifications are needed.

**Comment by pavlenex** [comment]
@Shourya742 can you fix this https://github.com/stratum-mining/stratum/actions/runs/10062714214/job/27906911880 ? Everything else is passing.

**Comment by pavlenex** [comment]
This one is waiting for @lorbax's review according to the last dev call and then it's good to go.

**Comment by lorbax** [comment]
I think that now the logic of the flag checking is fine.
Nevertheless, perhaps the author should open an issue regarding the fact that the standard behavior is to have the JDS work with that flag unset, and `async_mining_allowed` should offered optionally. Not having this may expose the pool to this [this](https://github.com/stratum-mining/stratum/issues/1053) attack (pointed out in [this](https://github.com/stratum-mining/stratum/issues/1053#issuecomment-2230607082)  comment).

File: roles/jd-server/config-examples/jds-config-local-example.toml
Comment (plebhash) [inline roles/jd-server/config-examples/jds-config-local-example.toml:2]: as suggested by @lorbax :https://github.com/stratum-mining/stratum/pull/1035#issuecomment-2252334208

> the standard behavior is to have the JDS work with that flag unset, and async_mining_allowed should offered optionally. Not having this may expose the pool to this https://github.com/stratum-mining/stratum/issues/1053 attack (pointed out in https://github.com/stratum-mining/stratum/issues/1053#issuecomment-2230607082 comment).

```suggestion
async_mining_allowed = false
```

**Comment by Fi3** [comment]
> I think that now the logic of the flag checking is fine. Nevertheless, perhaps the author should open an issue regarding the fact that the standard behavior is to have the JDS work with that flag unset, and `async_mining_allowed` should offered optionally. Not having this may expose the pool to this [this](https://github.com/stratum-mining/stratum/issues/1053) attack (pointed out in [this](https://github.com/stratum-mining/stratum/issues/1053#issuecomment-2230607082) comment).

if the flag is set this is not an attack but the desired behavior (by flag definition). For example if you save everything and you calculate the miner reward in a later moment or somewhere else and not each time that you receive a share (what I expect most pool will do) setting this flag is ok.

**Comment by lorbax** [comment]
> > I think that now the logic of the flag checking is fine. Nevertheless, perhaps the author should open an issue regarding the fact that the standard behavior is to have the JDS work with that flag unset, and `async_mining_allowed` should offered optionally. Not having this may expose the pool to this [this](https://github.com/stratum-mining/stratum/issues/1053) attack (pointed out in [this](https://github.com/stratum-mining/stratum/issues/1053#issuecomment-2230607082) comment).
> 
> if the flag is set this is not an attack but the desired behavior (by flag definition). For example if you save everything and you calculate the miner reward in a later moment or somewhere else and not each time that you receive a share (what I expect most pool will do) setting this flag is ok.
> 
> Also I do not thing that the issue that you refer should be considered an issue see my comment: [#1053 (comment)](https://github.com/stratum-mining/stratum/issues/1053#issuecomment-2238880037)

I agree that the issue does not exists, I misread it.
Nevertheless, JDS works only with that flag set, while that flag set indicates a feature.

**Comment by plebhash** [comment]
> Nevertheless, JDS works only with that flag set, while that flag set indicates a feature.

hopefully this PR fixes this

@lorbax do you think it's ready?

**Pull request merged**
